lib/rust/flexer/src/lib.rs

```diff
@@ -14,26 +14,239 @@
+        self.state_stack.pop()
+    }
+
+    /// End states until the specified `state` is reached, leaving the lexer in `state`.
+    ///
+    /// If `state` does not exist on the lexer's stack, then the lexer will be left in the root
+    /// state.
+    pub fn pop_states_until(&mut self, state:group::Identifier) -> Vec<group::Identifier> {
+        let non_opt_root_state_position =
+            self.state_stack.iter().positions(|elem| *elem == state).last().unwrap_or(0);
+        let range = (non_opt_root_state_position + 1)..self.state_stack.len();
+        self.state_stack.drain(range).collect()
+    }
+
+    /// Check if the lexer is currently in the state described by `state`.
+    pub fn is_in_state(&self, state:group::Identifier) -> bool {
+        self.current_state() == state
+    }
+}
+
+// === Trait Impls ===
+
+impl<Definition,Output,Reader> Deref for Flexer<Definition,Output,Reader> {
+    type Target = Definition;
+    fn deref(&self) -> &Self::Target {
+        &self.definition
+    }
+}
+
+impl<Definition,Output,Reader> DerefMut for Flexer<Definition,Output,Reader> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.definition
+    }
+}
+
+
+
+// ===================
+// === StageStatus ===
+// ===================
+
+/// The result of executing a single step of the DFA.
+#[derive(Clone,Copy,Debug,PartialEq)]
+pub enum StageStatus {
+    /// The initial state of a lexer stage.
+    Initial,
+    /// The stage exits successfully, having consumed a complete token.
+    ExitSuccess,
+    /// The stage exits unsuccessfully.
+    ExitFail,
+    /// A single step of the DFA has executed successfully.
+    ExitFinished,
+    /// The lexer should continue, transitioning to the included state.
+    ContinueWith(group::Identifier)
+}
+
+impl StageStatus {
+    /// Check if the lexer stage should continue.
+    pub fn should_continue(&self) -> bool {
+        self.continue_as().is_some()
+    }
+
+    /// Obtain the state to which the lexer should transition, iff the lexer should continue.
+    pub fn continue_as(&self) -> Option<group::Identifier> {
+        match self {
+            StageStatus::Initial           => Some(group::Identifier::new(0)),
+            StageStatus::ContinueWith(val) => Some(*val),
+            _                              => None
+        }
+    }
+}
+
+
+// === Trait Impls ===
+
+impl Default for StageStatus {
+    fn default() -> Self {
+        StageStatus::Initial
+    }
+}
+
+
+
+// ==============
+// === Result ===
+// ==============
+
+/// The result of executing the lexer on a given input.
+#[derive(Clone,Debug)]
+pub enum Result<T> {
+    /// The lexer succeeded, returning the contained token stream.
+    Success(Vec<T>),
+    /// The lexer succeeded on part of the input, returning the contained token stream.
+    Partial(Vec<T>),
+    /// The lexer failed on the input, returning any tokens it _did_ manage to consume.
+    Failure(Vec<T>)
```

Comment:
A thought - from a structural point of view, it would be nicer to keep `Vec<T>` as separate field and "status" `{Success,PArtial,Failure}` as other field. This way you can access tokens without matching etc. But its so small thing Im completely ok with leaving it as is